tests/utils/test_safedir_anchored.py

```
@@ -0,0 +1,203 @@
+    def test_parent_traversal_rejected(self, tmp_path: Path) -> None:
+        """``..`` components would escape — must be refused before any open."""
+        (tmp_path / "child").mkdir()
+        (tmp_path / "doc.txt").write_text("data")
+        with SafeDir.open_root(tmp_path / "child") as root:
+            with pytest.raises(ValueError, match=r"\.\."):
+                root.open_anchored_reader(Path("../doc.txt"))
+
+    def test_empty_path_rejected(self, tmp_path: Path) -> None:
+        """Empty relative_path doesn't identify any file."""
+        with SafeDir.open_root(tmp_path) as root:
+            with pytest.raises(ValueError):
```

Ruff (0.15.13)

[warning] `pytest.raises(ValueError)` is too broad, set the `match` parameter or use a more specific exception

(PT011)